### PR TITLE
Field path without annotation

### PR DESCRIPTION
**Pull request by ozangunalp**
The codegen uses Path annotations for associating fieldIds with accessors on the domain model.
This may not be ideal if the domain model is itself a result of generated code.

This WIP proposes a builder pattern based on method references to construct field paths external from the domain model.

The code generator uses field paths as alternative to Path annotations for generating FieldInfos and Wrapper classes.

### CODE REVIEW

**Comment by dubreuia** [comment]
This is necessary to generate code on a domain model you cannot edit (typically when it's itself a result of a generation). The first phase of "without annotations" could be itself generated (at least partially), it would serve as a pre-generation phase.

Just to recap:

**Current process**

- `SampleModel` annotated with `@Path` annotations
- Code generation with `doov-generator`, based on annotated model
- Generates `SampleFieldIdInfo`

**Without annotations**

- `SampleModel`, no annotation because you can't, then write `SampleFieldPath` with builder pattern
- Code generation with `doov-generator`, based on `SampleFieldPath`
- Generates `SampleFieldIdInfo`

**Comment by dubreuia** [comment]
Not a fan of the root sample2 folder. I'd prefer putting everything we don't release in the sample folder

- sample/sample1
  - sample/sample1/base
  - sample/sample1/generated
- sample/sample2
  - sample/sample2/base
  - sample/sample2/generated
- sample/example
- ...

File: generator/src/main/java/io/doov/gen/ModelMapGenMojo.java [hunk 1]
```diff
@@ -69,6 +75,12 @@
     @Parameter(required = true)
     private String packageFilter;
 
+    /**
+     * TODO find a better way to configure field path provider regarding fieldClasses & sourceClasses
+     */
+    @Parameter
+    private String fieldPathProvider;
```
Comment (dubreuia) [inline generator/src/main/java/io/doov/gen/ModelMapGenMojo.java]: I think changing this `String` to a `List<String>` and make it work like `fieldClasses` would be enough. 

If you need to have some `sourceClasses` with a `fieldPathProvider` and not for some, you can always split it into 2 maven executions

**Comment by ozangunalp** [comment]
Need to rebase this branch onto the new directory structure on master.

File: generator/src/main/java/io/doov/gen/ModelMapGenMojo.java (hunk 1 above)
Comment (ozangunalp) [reply generator/src/main/java/io/doov/gen/ModelMapGenMojo.java]: As-is `sourceClasses` and `fieldClasses` lists match each other using index order in the list.
A better way is to have a structure of configuration like 
```
<configuration>
    <packageFilter></packageFilter>
    <fieldClass></fieldClass>
    <sourceClass></sourceClass>
    <fieldPathProvider></fieldPathProvider>
</configuration>
```
`fieldPathProvider` being optional.

As you said different code generations can be split into separate executions.

**Comment by dubreuia** [comment]
Yeah sorry I moved everything, you can do it now, I'm finished moving stuff

File: generator/src/main/java/io/doov/gen/ModelMapGenMojo.java (hunk 1 above)
Comment (dubreuia) [reply generator/src/main/java/io/doov/gen/ModelMapGenMojo.java]: Yes your configuration is better, and `n` executions if needed

**Comment by dubreuia** [comment]
I'll fix the conflict and merge

**Comment by dubreuia** [comment]
Merged

**Pull request merged**